app/knights/prepare_to_battle.py

```diff
@@ -0,0 +1,26 @@
+from app.knights.basic_stats import Knight
+
+
+class KnightsPrepareToBattle:
+
+    def __init__(self, knight: Knight) -> None:
+        self.name = knight.name
+        self.power = knight.power
+        self.hp = knight.hp
+        self.protection = 0
+
+    def apply_armour(self, knight: Knight) -> None:
+        for element in knight.armour:
+            self.protection += element["protection"]
+
+    def apply_weapon(self, knight: Knight) -> None:
+        self.power += knight.weapon["power"]
+
+    def apply_potion(self, knight: Knight) -> None:
+        if knight.potion is not None:
+            if "power" in knight.potion["effect"]:
+                self.power += knight.potion["effect"]["power"]
+            if "hp" in knight.potion["effect"]:
+                self.hp += knight.potion["effect"]["hp"]
+            if "protection" in knight.potion["effect"]:
+                self.protection += knight.potion["effect"]["protection"]
```

Comment:
DRY. create a loop for these actions

Reply:
In this function, `def apply_potion(self, knight: Knight)`, I use just one knight as argument. And every knight has only one key — potion.
Then, in ready_to_battle.py I use the loop for applying additional elements.

```
    for knight_name, knight_stats in knights_instances.items():
        final_knights_stats[knight_name] = KnightsPrepareToBattle(knight_stats)
        final_knights_stats[knight_name].apply_armour(knight_stats)
        final_knights_stats[knight_name].apply_weapon(knight_stats)
        final_knights_stats[knight_name].apply_potion(knight_stats)
```

Reply:
Alex means that you perform the same actions for `"power", "hp", "protection"` so you could rewrite it like this:
```
if knight.potion is not None:
    for stat in ("power", "hp", "protection"):
         ...
```

Reply:
@IvanRamyk Thank you for explaining.
In your opinion, do I need to change logic and create attribute with dictionary for class KnightsPrepareToBattle?

Because I tried to find how to substitute values for `self.` and failed.

Another option — to make with for loop and check potion effect a bit differently. 

```
    def apply_potion(self, knight: Knight) -> None:
        if knight.potion is not None:
            for potion_effect in knight.potion["effect"]:
                if potion_effect == "power":
                    self.power += knight.potion["effect"][potion_effect]
                if potion_effect == "hp":
                    self.hp += knight.potion["effect"][potion_effect]
                if potion_effect == "protection":
                    self.protection += knight.potion["effect"][potion_effect]
```

Thank you!

Reply:
Other option is by `__setattr__` and `__getattribute__`

```
    def apply_potion(self, knight: Knight) -> None:
        if knight.potion is not None:
            for potion_effect in knight.potion["effect"]:
                self.__setattr__(
                    potion_effect,
                    self.__getattribute__(potion_effect)
                    + knight.potion["effect"][potion_effect])
```

Reply:
The last option is almost fine, but you can use `setattr` and `getattr` functions and pass self as the first argument (IMHO, it looks more readable than magic methods)

Reply:
Thank you so much for your help!